storagenode/storagenodedb/bandwidthdb.go

```diff
@@ -220,6 +267,110 @@ func (db *bandwidthDB) Rollup(ctx context.Context) (err error) {
 	return nil
 }
 
+// GetDailyRollups returns slice of daily bandwidth usage rollups for provided time range,
+// sorted in ascending order
+func (db *bandwidthDB) GetDailyRollups(ctx context.Context, from, to time.Time) (_ []bandwidth.UsageRollup, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	since, _ := date.DayBoundary(from.UTC())
+	_, before := date.DayBoundary(to.UTC())
+
+	return db.getDailyUsageRollups(ctx,
+		"WHERE DATETIME(?) <= DATETIME(interval_start) AND DATETIME(interval_start) <= DATETIME(?)",
+		since, before,
+		since, before)
+}
+
+// GetDailySatelliteRollups returns slice of daily bandwidth usage for provided time range,
+// sorted in ascending order for particular satellite
+func (db *bandwidthDB) GetDailySatelliteRollups(ctx context.Context, satelliteID storj.NodeID, from, to time.Time) (_ []bandwidth.UsageRollup, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	since, _ := date.DayBoundary(from.UTC())
+	_, before := date.DayBoundary(to.UTC())
+
+	return db.getDailyUsageRollups(ctx,
+		"WHERE satellite_id = ? AND DATETIME(?) <= DATETIME(interval_start) AND DATETIME(interval_start) <= DATETIME(?)",
+		satelliteID, since, before,
+		satelliteID, since, before)
+}
+
+// getDailyUsageRollups returns slice of grouped by date bandwidth usage rollups
+// sorted in ascending order and applied condition if any
+func (db *bandwidthDB) getDailyUsageRollups(ctx context.Context, cond string, args ...interface{}) (_ []bandwidth.UsageRollup, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	query := `SELECT action, sum(a) as amount, DATETIME(DATE(interval_start)) as date FROM (
+			SELECT action, sum(amount) as a, created_at AS interval_start
+			FROM bandwidth_usage
+			` + cond + `
+			GROUP BY interval_start, action
+				UNION ALL
+					SELECT action, sum(amount) as a, interval_start
+					FROM bandwidth_usage_rollups
+					` + cond + `
+					GROUP BY interval_start, action
+		) GROUP BY date, action
+		ORDER BY interval_start`
+
+	rows, err := db.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err = errs.Combine(err, rows.Close())
+	}()
+
+	var dates []time.Time
+	usageRollupsByDate := make(map[time.Time]*bandwidth.UsageRollup)
+
+	for rows.Next() {
+		var action int32
+		var amount int64
+		var intervalStartN dbutil.NullTime
+
+		err = rows.Scan(&action, &amount, &intervalStartN)
+		if err != nil {
+			return nil, err
+		}
+
+		intervalStart := intervalStartN.Time
+
+		rollup, ok := usageRollupsByDate[intervalStart]
+		if !ok {
```

Comment:
Can rollup be accessed more than once?

Reply:
Yes, for each `action`

Reply:
Then, should the amounts in https://github.com/storj/storj/pull/2923/files/7c76beaf56d6c7749446fa5aaa770b662318ae5f#diff-4caa485c1945191cb12aa56ab08930fdR350 be added (i.e. `+=`) rather than assigned (i.e. `=`)?

Reply:
no, each action is rolled up, so for one day you get `len(rows)` equals `len(actions)`

Reply:
Thanks for the explanation

Reply:
you are welcome!)